File: KwangHo/올바른 괄호.py
def solution(s):
    answer = True
    dap=[]
    
    if s[0]==")": # 첫값이 )라면 계신 불필요해서 리턴
        return False
    
    for i in s :
        if i == "(" :  # ( 는 추가
            dap.append(i)
        elif i == ")" and len(dap)>0 and dap[len(dap)-1] == "(" : # ) 는 스택이 0보다 크거나 + ( 라면 스택 팝
            dap.pop()
        else : 
            return False # 위 조건이 아니라면 리턴
            break
    
    if len(dap) > 0 : # 남아있는게 하나라도 있으면 False
        answer = False

    return answer

'''
열고닫기 한세트 구성이라 스택으로 차례대로
집어넣다가 세트나오면 없애주는 방식으로 하고
마지막에 스택에 0이면 올바른 괄호 !!
+ 중간중간 불필요한 계산으로 인한 종료점 추가
'''

File: Doodream/올바른 괄호.js
function solution(s) {
    var answer = true;
    let sArr = s.split('');

    // 주어진 s가 짝수가 아니거나 처음이 '(' 마지막이 ')' 로 끝나지 않는 경우 x
    if (s.length % 2 !== 0 || sArr[0] !== '(' || sArr[s.length - 1] !== ')') {
        answer = false;
        return answer;
    }

    // '(' ')' 의 갯수가 동일하지 않은경우 x
    let openLetterCount = 0;
    let closeLetterCount = 0;

    sArr.map(x => {
        x === '(' ? openLetterCount++ : closeLetterCount++;
    })

    if (openLetterCount !== closeLetterCount) {
        answer = false;
        return answer;
    }

    // 문자열에서 괄호가 닫힌 경우를 모두 제거하고 (반복)제거 후 문자열로 통합
    sStr = s.split('()').join('');
    // 해당 문자열이 () 혹은 없는 경우 true
    if (sStr === '()' || sStr === '') {
        return answer;
    }

    // false 인 경우는 '()))((()' 이런 경우 이므로 바로 위코드를 수행하면 ')))(((' 이렇게 되어 거른다. 
    if (sStr[0] !== '(' || sStr[sStr.length - 1] !== ')') {
        answer = false;
        return answer;
    }

    return answer;
}

let s = "()))((()";
console.log(solution(s));